src/Sarif.Viewer.VisualStudio/CodeAnalysisResultManager.cs

```diff
@@ -637,5 +611,220 @@ internal bool TryResolveFilePathFromSolution(string solutionPath, string pathFro
+                dataCache.RemappedUriBasePaths[uriBaseId] = new Uri(resolvedPath.Substring(0, uriBaseIdEndIndex), UriKind.Absolute);
+            }
+            else
+            {
+                // If there's no relativeUri/uriBaseId pair or we couldn't determine the uriBaseId value,
+                // map the original prefix to the new prefix.
+                dataCache.RemappedPathPrefixes.Add(new Tuple<string, string>(originalPrefix, resolvedPrefix));
+            }
+
+            return true;
+        }
+
+        // return false means cannot resolve local file and will use embedded file.
+        internal bool VerifyFileWithArtifactHash(SarifErrorListItem sarifErrorListItem, string pathFromLogFile, RunDataCache dataCache, string resolvedPath, string embeddedTempFilePath, out string newResolvedPath)
+        {
+            newResolvedPath = null;
+
+            if (string.IsNullOrEmpty(resolvedPath))
+            {
+                // cannot find corresponding file in local, then use embedded file
+                newResolvedPath = embeddedTempFilePath;
+                return true;
+            }
+
+            if (!dataCache.FileDetails.ContainsKey(pathFromLogFile))
+            {
+                newResolvedPath = resolvedPath;
+                return true;
+            }
+
+            ArtifactDetailsModel fileData = dataCache.FileDetails[pathFromLogFile];
+            string fileHash = this.GetFileHash(this._fileSystem, resolvedPath);
+
+            if (fileHash.Equals(fileData.Sha256Hash))
+            {
+                // found a file in file system which has same hashcode as embeded content.
+                newResolvedPath = resolvedPath;
+                return true;
+            }
+
+            bool hasEmbeddedContent = !string.IsNullOrEmpty(embeddedTempFilePath);
+            ResolveEmbeddedFileDialogResult dialogResult;
+            if (userDialogPreference.TryGetValue(sarifErrorListItem.LogFilePath, out ResolveEmbeddedFileDialogResult preference) &&
+                preference != ResolveEmbeddedFileDialogResult.None &&
+
+                // if preference is OpenEmbeddedFileContent but this result has no embedded content, should ignore this preference
+                !(!hasEmbeddedContent && preference == ResolveEmbeddedFileDialogResult.OpenEmbeddedFileContent))
+            {
+                dialogResult = preference;
+            }
+            else
+            {
+                var dialog = new ResolveEmbeddedFileDialog(hasEmbeddedContent);
+                dialog.ShowModal();
+                dialogResult = dialog.Result;
+                if (dialog.ApplyUserPreference)
+                {
+                    userDialogPreference.AddOrUpdate(sarifErrorListItem.LogFilePath, dialogResult, (key, value) => dialogResult);
+                }
+            }
+
+            if (dialogResult == ResolveEmbeddedFileDialogResult.None)
```

Comment:
>if (dialogResult [](start = 12, length = 16)

create a switch. and you will probably prevent the last return

Reply:
done

---
In reply to: [606208909](https://github.com/microsoft/sarif-visualstudio-extension/pull/384#discussion_r606208909) [](ancestors = 606208909)